File: app/src/main/java/com/guardcall/app/engine/SkepticismEngine.kt
package com.guardcall.app.engine

import android.content.Context
import com.guardcall.app.data.AppDatabase
import com.guardcall.app.data.OfficialProcedure
import com.guardcall.app.data.ScamPattern
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.runBlocking

class SkepticismEngine {

    private val engineScope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    data class AnalysisResult(
        val riskScore: Int,
        val advice: String,
        val contradictionFound: String?,
        val warningText: String?,
        val detectedPayments: List<String> = emptyList()
    )

    fun analyzeDialogue(context: Context, text: String): AnalysisResult {
        // Query database synchronously for analysis since this is called on-demand by accessibility service thread
        val database = AppDatabase.getDatabase(context.applicationContext, engineScope)
        
        var calculatedRisk = 0
        val advices = mutableListOf<String>()
        var contradiction: String? = null
        var warningText: String? = null

        // Detect protected payment methods
        val detectedPayments = detectTargetedPayments(context, text)
        if (detectedPayments.isNotEmpty()) {
            calculatedRisk += 15 * detectedPayments.size
            advices.add("Payment Channel Threatened: ${detectedPayments.joinToString(", ")}")
        }

        runBlocking(Dispatchers.IO) {
            val procedures = database.scamDao().getAllProceduresList()
            val patterns = database.scamDao().getAllPatternsList()

            // 1. Scan for Scam Patterns / Keywords (News & Police alerts)
            for (pattern in patterns) {
                val keywordsList = pattern.keywords.split(",")
                var matchCount = 0
                for (kw in keywordsList) {
                    val trimmed = kw.trim()
                    if (trimmed.isNotEmpty() && text.contains(trimmed, ignoreCase = true)) {
                        matchCount++
                    }
                }
                
                if (matchCount > 0) {
                    calculatedRisk += 25 * matchCount
                    advices.add(pattern.advice)
                }
            }

            // 2. Scan for Procedure Inconsistencies (What official organizations NEVER do)
            // e.g. Spammer says: "郵寄你張銀行卡" (mail your bank card)
            // Procedure says: Police will "never ask you to mail physical bank cards"
            for (proc in procedures) {
                // If text contains the forbidden actions or keywords associated with the institution
                val matchesInstitution = containsInstitutionRef(text, proc.institution)
                val matchesForbiddenAction = containsForbiddenActionKeywords(text, proc.forbiddenAction)

                if (matchesInstitution && matchesForbiddenAction) {
                    calculatedRisk += 60
                    contradiction = "${proc.institution} contradiction: claimed procedure involves ${proc.forbiddenAction}."
                    warningText = proc.warningText
                    advices.add("Inconsistency Detected: ${proc.warningText}")
                }
            }
        }

        // Cap risk score at 100%
        val finalRiskScore = calculatedRisk.coerceAtMost(100)
        val finalAdvice = if (advices.isNotEmpty()) advices.distinct().joinToString("\n") else "No immediate contradictions found. Keep listening."

        return AnalysisResult(
            riskScore = finalRiskScore,
            advice = finalAdvice,
            contradictionFound = contradiction,
            warningText = warningText,
            detectedPayments = detectedPayments
        )
    }

    fun detectTargetedPayments(context: Context, text: String): List<String> {
        val sharedPrefs = context.getSharedPreferences("GuardCallPrefs", Context.MODE_PRIVATE)
        // Get user-protected payment methods set (default to all of them enabled)
        val protectedPayments = sharedPrefs.getStringSet("protected_payments", setOf("Bank Apps", "Internet Banking", "Visa", "FPS", "Octopus", "7-Eleven", "Circle K", "Alipay", "WeChat Pay")) ?: emptySet()
        
        val detected = mutableListOf<String>()
        
        // Match conditions (in English and Cantonese)
        if (protectedPayments.contains("Bank Apps") && (text.contains("銀行app", ignoreCase = true) || text.contains("手機銀行", ignoreCase = true) || text.contains("bank app", ignoreCase = true))) {
            detected.add("Bank Apps")
        }
        if (protectedPayments.contains("Internet Banking") && (text.contains("網上銀行", ignoreCase = true) || text.contains("網銀", ignoreCase = true) || text.contains("e-banking", ignoreCase = true) || text.contains("internet banking", ignoreCase = true))) {
            detected.add("Internet Banking")
        }
        if (protectedPayments.contains("Visa") && (text.contains("visa", ignoreCase = true) || text.contains("信用卡", ignoreCase = true) || text.contains("credit card", ignoreCase = true))) {
            detected.add("Visa")
        }
        if (protectedPayments.contains("FPS") && (text.contains("轉數快", ignoreCase = true) || text.contains("fps", ignoreCase = true) || text.contains("快速支付", ignoreCase = true))) {
            detected.add("FPS")
        }
        if (protectedPayments.contains("Octopus") && (text.contains("八達通", ignoreCase = true) || text.contains("octopus", ignoreCase = true))) {
            detected.add("Octopus")
        }
        if (protectedPayments.contains("Alipay") && (text.contains("支付寶", ignoreCase = true) || text.contains("alipay", ignoreCase = true) || text.contains("支寶", ignoreCase = true))) {
            detected.add("Alipay")
        }
        if (protectedPayments.contains("WeChat Pay") && (text.contains("微信支付", ignoreCase = true) || text.contains("wechat pay", ignoreCase = true) || text.contains("微信過數", ignoreCase = true) || text.contains("微訊", ignoreCase = true))) {
            detected.add("WeChat Pay")
        }
        if (protectedPayments.contains("7-Eleven") && (text.contains("7-11", ignoreCase = true) || text.contains("7-eleven", ignoreCase = true) || text.contains("便利店", ignoreCase = true) || text.contains("七十一", ignoreCase = true))) {
            detected.add("7-Eleven")
        }
        if (protectedPayments.contains("Circle K") && (text.contains("circle k", ignoreCase = true) || text.contains("ok便利店", ignoreCase = true) || text.contains("ok 點", ignoreCase = true) || text.contains("ok便利", ignoreCase = true))) {
            detected.add("Circle K")
        }
        
        return detected
    }

    private fun containsInstitutionRef(text: String, institutionName: String): Boolean {
        // Simple heuristic matching for institution names in HK/English
        val aliases = when {
            institutionName.contains("Police", ignoreCase = true) -> listOf("Police", "警察", "警局", "公安", "反詐")
            institutionName.contains("Alipay", ignoreCase = true) -> listOf("Alipay", "支付寶", "支寶")
            institutionName.contains("HSBC", ignoreCase = true) -> listOf("HSBC", "滙豐", "匯豐", "銀行")
            else -> listOf(institutionName)
        }
        return aliases.any { text.contains(it, ignoreCase = true) }
    }

    private fun containsForbiddenActionKeywords(text: String, forbiddenAction: String): Boolean {
        // Map common forbidden actions to Cantonese/English keywords
        val keywords = when {
            forbiddenAction.contains("mail", ignoreCase = true) || forbiddenAction.contains("card", ignoreCase = true) -> 
                listOf("郵寄", "寄過嚟", "寄黎", "銀行卡", "提款卡", "密碼", "mail", "post", "card")
            forbiddenAction.contains("transfer", ignoreCase = true) || forbiddenAction.contains("account", ignoreCase = true) -> 
                listOf("轉賬", "安全賬戶", "安全賬戶", "匯款", "過數", "轉帳", "transfer", "account", "fund")
            forbiddenAction.contains("security code", ignoreCase = true) || forbiddenAction.contains("verification", ignoreCase = true) -> 
                listOf("驗證碼", "安全編碼", "一次性密碼", "安全碼", "verification", "otp", "code")
            else -> forbiddenAction.split(" ")
        }
        return keywords.any { text.contains(it, ignoreCase = true) }
    }
}

File: ios/ScamCall/SkepticismEngine.swift
import Foundation

class SkepticismEngine {
    
    struct AnalysisResult {
        let riskScore: Int
        let advice: String
        let contradictionFound: String?
        let warningText: String?
        let detectedPayments: [String]
    }
    
    func analyzeDialogue(text: String, procedures: [OfficialProcedure], patterns: [ScamPattern]) -> AnalysisResult {
        var calculatedRisk = 0
        var advices: [String] = []
        var contradiction: String? = nil
        var warningText: String? = nil
        
        // 1. Detect protected payment methods
        var detectedPayments: [String] = []
        let defaults = UserDefaults.standard
        
        let checkBankApps = defaults.object(forKey: "protect_bank_apps") == nil ? true : defaults.bool(forKey: "protect_bank_apps")
        let checkInternetBanking = defaults.object(forKey: "protect_internet_banking") == nil ? true : defaults.bool(forKey: "protect_internet_banking")
        let checkVisa = defaults.object(forKey: "protect_visa") == nil ? true : defaults.bool(forKey: "protect_visa")
        let checkFps = defaults.object(forKey: "protect_fps") == nil ? true : defaults.bool(forKey: "protect_fps")
        let checkOctopus = defaults.object(forKey: "protect_octopus") == nil ? true : defaults.bool(forKey: "protect_octopus")
        let checkAlipay = defaults.object(forKey: "protect_alipay") == nil ? true : defaults.bool(forKey: "protect_alipay")
        let checkWeChatPay = defaults.object(forKey: "protect_wechat_pay") == nil ? true : defaults.bool(forKey: "protect_wechat_pay")
        let check7Eleven = defaults.object(forKey: "protect_7_eleven") == nil ? true : defaults.bool(forKey: "protect_7_eleven")
        let checkCircleK = defaults.object(forKey: "protect_circle_k") == nil ? true : defaults.bool(forKey: "protect_circle_k")
        
        if checkBankApps && (text.localizedCaseInsensitiveContains("銀行app") || text.localizedCaseInsensitiveContains("手機銀行") || text.localizedCaseInsensitiveContains("bank app")) {
            detectedPayments.append("Bank Apps")
        }
        if checkInternetBanking && (text.localizedCaseInsensitiveContains("網上銀行") || text.localizedCaseInsensitiveContains("網銀") || text.localizedCaseInsensitiveContains("e-banking") || text.localizedCaseInsensitiveContains("internet banking")) {
            detectedPayments.append("Internet Banking")
        }
        if checkVisa && (text.localizedCaseInsensitiveContains("visa") || text.localizedCaseInsensitiveContains("信用卡") || text.localizedCaseInsensitiveContains("credit card")) {
            detectedPayments.append("Visa")
        }
        if checkFps && (text.localizedCaseInsensitiveContains("轉數快") || text.localizedCaseInsensitiveContains("fps") || text.localizedCaseInsensitiveContains("快速支付")) {
            detectedPayments.append("FPS")
        }
        if checkOctopus && (text.localizedCaseInsensitiveContains("八達通") || text.localizedCaseInsensitiveContains("octopus")) {
            detectedPayments.append("Octopus")
        }
        if checkAlipay && (text.localizedCaseInsensitiveContains("支付寶") || text.localizedCaseInsensitiveContains("alipay") || text.localizedCaseInsensitiveContains("支寶")) {
            detectedPayments.append("Alipay")
        }
        if checkWeChatPay && (text.localizedCaseInsensitiveContains("微信支付") || text.localizedCaseInsensitiveContains("wechat pay") || text.localizedCaseInsensitiveContains("微信過數") || text.localizedCaseInsensitiveContains("微訊")) {
            detectedPayments.append("WeChat Pay")
        }
        if check7Eleven && (text.localizedCaseInsensitiveContains("7-11") || text.localizedCaseInsensitiveContains("7-eleven") || text.localizedCaseInsensitiveContains("便利店") || text.localizedCaseInsensitiveContains("七十一")) {
            detectedPayments.append("7-Eleven")
        }
        if checkCircleK && (text.localizedCaseInsensitiveContains("circle k") || text.localizedCaseInsensitiveContains("ok便利店") || text.localizedCaseInsensitiveContains("ok 點") || text.localizedCaseInsensitiveContains("ok便利")) {
            detectedPayments.append("Circle K")
        }
        
        if !detectedPayments.isEmpty {
            calculatedRisk += 15 * detectedPayments.count
            advices.append("Payment Channel Threatened: \(detectedPayments.joined(separator: ", "))")
        }
        
        // 2. Scan for Scam Patterns / Keywords
        for pattern in patterns {
            let keywordsList = pattern.keywords.components(separatedBy: ",")
            var matchCount = 0
            for kw in keywordsList {
                let trimmed = kw.trimmingCharacters(in: .whitespacesAndNewlines)
                if !trimmed.isEmpty && text.localizedCaseInsensitiveContains(trimmed) {
                    matchCount += 1
                }
            }
            
            if matchCount > 0 {
                calculatedRisk += 25 * matchCount
                advices.append(pattern.advice)
            }
        }
        
        // 3. Scan for Procedure Inconsistencies
        for proc in procedures {
            let matchesInstitution = containsInstitutionRef(text: text, institutionName: proc.institution)
            let matchesForbiddenAction = containsForbiddenActionKeywords(text: text, forbiddenAction: proc.forbiddenAction)
            
            if matchesInstitution && matchesForbiddenAction {
                calculatedRisk += 60
                contradiction = "\(proc.institution) contradiction: claimed procedure involves \(proc.forbiddenAction)."
                warningText = proc.warningText
                advices.append("Inconsistency Detected: \(proc.warningText)")
            }
        }
        
        let finalRiskScore = min(calculatedRisk, 100)
        let finalAdvice = advices.isEmpty ? "No immediate contradictions found. Keep listening." : Array(Set(advices)).joined(separator: "\n")
        
        return AnalysisResult(
            riskScore: finalRiskScore,
            advice: finalAdvice,
            contradictionFound: contradiction,
            warningText: warningText,
            detectedPayments: detectedPayments
        )
    }
    
    private func containsInstitutionRef(text: String, institutionName: String) -> Bool {
        let aliases: [String]
        if institutionName.localizedCaseInsensitiveContains("Police") || institutionName.localizedCaseInsensitiveContains("警察") {
            aliases = ["Police", "警察", "警局", "公安", "反詐"]
        } else if institutionName.localizedCaseInsensitiveContains("Alipay") || institutionName.localizedCaseInsensitiveContains("支付寶") {
            aliases = ["Alipay", "支付寶", "支寶"]
        } else if institutionName.localizedCaseInsensitiveContains("HSBC") || institutionName.localizedCaseInsensitiveContains("滙豐") {
            aliases = ["HSBC", "滙豐", "匯豐", "銀行"]
        } else {
            aliases = [institutionName]
        }
        
        return aliases.contains { text.localizedCaseInsensitiveContains($0) }
    }
    
    private func containsForbiddenActionKeywords(text: String, forbiddenAction: String) -> Bool {
        let keywords: [String]
        if forbiddenAction.localizedCaseInsensitiveContains("mail") || forbiddenAction.localizedCaseInsensitiveContains("card") || forbiddenAction.localizedCaseInsensitiveContains("郵寄") {
            keywords = ["郵寄", "寄過嚟", "寄黎", "銀行卡", "提款卡", "密碼", "mail", "post", "card"]
        } else if forbiddenAction.localizedCaseInsensitiveContains("transfer") || forbiddenAction.localizedCaseInsensitiveContains("account") || forbiddenAction.localizedCaseInsensitiveContains("轉賬") {
            keywords = ["轉賬", "安全賬戶", "安全賬戶", "匯款", "過數", "轉帳", "transfer", "account", "fund"]
        } else if forbiddenAction.localizedCaseInsensitiveContains("security code") || forbiddenAction.localizedCaseInsensitiveContains("verification") || forbiddenAction.localizedCaseInsensitiveContains("驗證碼") {
            keywords = ["驗證碼", "安全編碼", "一次性密碼", "安全碼", "verification", "otp", "code"]
        } else {
            keywords = forbiddenAction.components(separatedBy: " ")
        }
        
        return keywords.contains { text.localizedCaseInsensitiveContains($0) }
    }
}
